cranelift-codegen/src/regalloc/branch_splitting.rs

```diff
@@ -0,0 +1,223 @@
+                            .brff(cond, val, new_ebb, &[]);
+                    }
+                }
+                _ => {
+                    panic!("Unhandled side exit type");
+                }
+            }
+            let ok = self.cur.func.update_encoding(inst, self.cur.isa).is_ok();
+            debug_assert!(ok);
+
+            // Insert a jump to the original target with the original arguments into the new block.
+            self.cur.goto_first_insertion_point(new_ebb);
+            self.cur.ins().jump(target, jump_args.as_slice());
+
+            // Reset the cursor to point to the branch.
+            self.cur.goto_inst(inst);
+        }
+    }
+
+    fn visit_terminator_branch(&mut self, inst: Inst) {
+        let inst_data = &self.cur.func.dfg[inst];
+        let opcode = inst_data.opcode();
+        if opcode != Opcode::Jump && opcode != Opcode::Fallthrough {
+            // This opcode is ignored as it does not have any EBB parameters.
+            if opcode != Opcode::IndirectJumpTableBr {
+                debug_assert!(!opcode.is_branch())
+            }
+            return;
+        }
+
+        let target = match inst_data {
+            InstructionData::Jump { destination, .. } => destination,
+            _ => panic!(
+                "Unexpected instruction {} in visit_terminator_branch",
+                self.cur.display_inst(inst)
+            ),
+        };
+        debug_assert!(self.cur.func.dfg[inst].opcode().is_terminator());
+
+        // If there are any parameters, split the edge.
+        if self.should_split_edge(inst, *target) {
+            // Create the block the branch will jump to.
+            let new_ebb = self.cur.func.dfg.make_ebb();
+            self.new_blocks = true;
+
+            // Split the current block before its terminator, and insert a new jump instruction to
+            // jump to it.
+            let jump = self.cur.ins().jump(new_ebb, &[]);
+            self.cur.insert_ebb(new_ebb);
+
+            // Reset the cursor to point to new terminator of the old ebb.
+            self.cur.goto_inst(jump);
+        }
+    }
+
+    // A new ebb must be inserted before the last ebb because the last ebb may have a
+    // fallthrough_return and can't have anything after it.
+    fn make_empty_ebb(&mut self) -> Ebb {
+        let new_ebb = self.cur.func.dfg.make_ebb();
+        let last_ebb = self.cur.layout().last_ebb().unwrap();
+        self.cur.layout_mut().insert_ebb(new_ebb, last_ebb);
+        self.new_blocks = true;
+        new_ebb
+    }
+
+    /// Returns whether we should introduce a new branch.
+    fn should_split_edge(&self, _branch: Inst, target: Ebb) -> bool {
+        // We should split the edge if the target has any parameters.
+        if self.cur.func.dfg.ebb_params(target).len() > 0 {
+            return true;
+        };
+
+        // Or, if the target is has more than one block reaching it.
+        if self.cfg.pred_iter(target).count() != 1 {
```

Comment:
`Iterator.count()` can be O(n), because it consumes the iterator. To check whether there is more than one element, it can do something like:

```rust
if let Some(_) = self.cfg.pred_iter(target).skip(1).next() {
````